lib/workers/repository/updates/generate.ts

```diff
@@ -355,6 +363,27 @@ export function generateBranchConfig(
     releaseTimestamp: releaseTimestamp!,
   }; // TODO: fixme (#9666)
 
+  let highestPrioritySemanticCommitType = '';
+
+  for (const upgrade of config.upgrades) {
+    if (
+      upgrade.semanticCommitType &&
+      upgrade.semanticCommitType in semanticCommitTypePriorities
+    ) {
+      const priority = semanticCommitTypePriorities[upgrade.semanticCommitType];
+      const highestPriority =
+        semanticCommitTypePriorities[highestPrioritySemanticCommitType] ?? -1;
+
+      if (priority > highestPriority) {
+        highestPrioritySemanticCommitType = upgrade.semanticCommitType;
+      }
+    }
+  }
+
+  if (highestPrioritySemanticCommitType) {
+    config.semanticCommitType = highestPrioritySemanticCommitType;
+  }
+
```

Comment:
```suggestion
config.semanticCommits = config.upgrades.some((upgrade) => upgrade.semanticCommits));

```

Reply:
Hey @rarkins, I applied this suggestion however later realised that `semanticCommits` is a union of strings, not a boolean:

<img width="743" alt="Screenshot 2024-10-22 at 10 22 05" src="https://github.com/user-attachments/assets/b8449902-7d0b-4d64-a525-4d4ace58e7bc">

What would you suggest we do here, check if some of them are `=== 'enabled' || === 'auto'`, and if so set it to `enabled`?

Reply:
Yeah, I forgot about that sorry. In short, while you're doing this, you should also make sure that "whether or not to use semantic commits in the first place" is also correctly handled if there's mixed values in upgrades in the same branch.

I'm not certain but this logic might be "Set config.semanticCommits to `enabled` if any upgrade has it set to `enabled`". I'm not sure that `auto` is still present at this point, needs to be checked.

Reply:
Right, from what I see here (if I've followed the code around correctly):

https://github.com/renovatebot/renovate/blob/c7d66d15af867623f6aee42c2a198085bfe59a16/lib/workers/repository/index.ts#L70-L79

The "top-level" auto can still be present at this point, since `detectSemanticCommits()` is called _after_ `extractDependencies()`. However now I'm not sure whether `generateBranchConfig()` should be taking into consideration these top-level config properties, or just the package/branch based options (?).

I'm wondering if the solution should be something like the following (pseudocode):

```
foreach config.upgrades

  if upgrade object contains "semanticCommits"

    if some of the upgrade objects contain "semanticCommits: enabled"
       set config.sematicCommits = "enabled"
       break;
    endif

    if some of the upgrade objects contain "semanticCommits: auto"
       set config.sematicCommits = await detectSemanticCommits()
    endif

  endif

end foreach
```

Reply:
It think you can assume that upgrade.semanticCommits is already set to `enabled` or `disabled`

Reply:
Ok, are you saying no further logic is necessary?

Or that we just need to check for `.some(u => u.semanticCommits === 'enabled')`?

Reply:
Yes, the latter. if any contains 'enabled' then make the branch enabled, otherwise disabled (which it should already be).

Reply:
Should be good now. I've added that logic plus a couple of minor code comments.